### ISSUE TITLE
avoid unnecessary Def objects

### ISSUE THREAD

**Issue by JonathanStarup** [flix/flix#2720]
When the backend wants to call a non-tail non-closure function it:
1. constructs a new Def object
2. PUTFIELD all the arguments
3. call unwind that then:
4. calls invoke on the def object that then:
5. GETFIELD and store the arguments

This means that the def object is only used to pass along the arguments. We could make another static invoke on the def class that receives the function arguments as jvm arguments directly.

The only problem I see is that self-recursive tail calls are right now implemented by reassigning the arg field and GOTO onto the start of the function, but there are no fields to reassign in this case. This can be solved by reassigning the local variables instead but it means the the generator needs to be aware of how it was called. Another issue is the a tail call is implemented by returning a contiuation, which would just be wrong here. instead this static invoke should trampoline the continuations by storing them in a local variable (combined with a goto probably to mimic returns normal control break).

So in conclusion:
* [ ] Add a static invoke function that has jvm arguments implements a trampoline
* [ ] modify applyDefTail, ApplyCloTail, and ApplySelfTail when in a static invoke call

P.S. @magnus-madsen should we also make a style.md format for issues?

**Comment by JonathanStarup** [comment]
put on hold for #2591

**Comment by magnus-madsen** [comment]
Tagging as future and closing. I think once we have algebraic effects we will have to rethink some of these issues.
